src/payment/data/sagas.test.js

```diff
@@ -863,6 +872,225 @@ describe('saga tests', () => {
     });
   });
 
+  describe('handlePaymentState', () => {
+    /** This function allows us to create a canned state */
+    const generateTestBasketState = (inPaymentState, inPayments = []) => ({
+      loading: true,
+      loaded: false,
+      submitting: false,
+      redirect: false,
+      isBasketProcessing: false,
+      basketId: '95472447-7B79-4F3D-A6BB-0231B5E3875B',
+      payments: inPayments,
+      paymentState: inPaymentState,
+      paymentStatePolling: {
+        keepPolling: false,
+        errorCount: DEFAULT_PAYMENT_STATE_POLLING_MAX_ERRORS,
+      },
+    });
+
+    /* Our various responses */
+    const expectQuickSuccess = () => ([
+      pollPaymentState.fulfill(),
+    ]);
+
+    const expectPollAndSucceed = () => ([
+      pollPaymentState.received({ state: PAYMENT_STATE.COMPLETED }),
+      pollPaymentState.fulfill(),
+    ]);
+
+    const expectRuntimeFailure = () => {
+      const basketMessageError = generateApiError([
+        {
+          error_code: ERROR_CODES.BASKET_CHANGED,
+          user_message: 'error',
+        },
+      ], false);
+
+      return [
+        pollPaymentState.failure(basketMessageError),
+        addMessage(ERROR_CODES.BASKET_CHANGED, 'error', {}, MESSAGE_TYPES.ERROR),
+      ];
+    };
+
+    const baseStateTable = Object.keys(PAYMENT_STATE)
```

Comment:
Here we loop through all our states and in a few actions build a testing plan for the different combinations or success and failure.